// @angular-devkit/core/src/terminal/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const caps = require("./caps");
const colors_1 = require("./colors");
exports.reset = caps.stdout.colors ? colors_1.colors.reset : (x) => x;
exports.bold = caps.stdout.colors ? colors_1.colors.bold : (x) => x;
exports.dim = caps.stdout.colors ? colors_1.colors.dim : (x) => x;
exports.italic = caps.stdout.colors ? colors_1.colors.italic : (x) => x;
exports.underline = caps.stdout.colors ? colors_1.colors.underline : (x) => x;
exports.inverse = caps.stdout.colors ? colors_1.colors.inverse : (x) => x;
exports.hidden = caps.stdout.colors ? colors_1.colors.hidden : (x) => x;
exports.strikethrough = caps.stdout.colors ? colors_1.colors.strikethrough : (x) => x;
exports.black = caps.stdout.colors ? colors_1.colors.black : (x) => x;
exports.red = caps.stdout.colors ? colors_1.colors.red : (x) => x;
exports.green = caps.stdout.colors ? colors_1.colors.green : (x) => x;
exports.yellow = caps.stdout.colors ? colors_1.colors.yellow : (x) => x;
exports.blue = caps.stdout.colors ? colors_1.colors.blue : (x) => x;
exports.magenta = caps.stdout.colors ? colors_1.colors.magenta : (x) => x;
exports.cyan = caps.stdout.colors ? colors_1.colors.cyan : (x) => x;
exports.white = caps.stdout.colors ? colors_1.colors.white : (x) => x;
exports.grey = caps.stdout.colors ? colors_1.colors.gray : (x) => x;
exports.gray = caps.stdout.colors ? colors_1.colors.gray : (x) => x;
exports.bgBlack = caps.stdout.colors ? colors_1.colors.bgBlack : (x) => x;
exports.bgRed = caps.stdout.colors ? colors_1.colors.bgRed : (x) => x;
exports.bgGreen = caps.stdout.colors ? colors_1.colors.bgGreen : (x) => x;
exports.bgYellow = caps.stdout.colors ? colors_1.colors.bgYellow : (x) => x;
exports.bgBlue = caps.stdout.colors ? colors_1.colors.bgBlue : (x) => x;
exports.bgMagenta = caps.stdout.colors ? colors_1.colors.bgMagenta : (x) => x;
exports.bgCyan = caps.stdout.colors ? colors_1.colors.bgCyan : (x) => x;
exports.bgWhite = caps.stdout.colors ? colors_1.colors.bgWhite : (x) => x;
